**logistic regression:**
regression analysis to conduct when the dependent is binary - predictive analysis.
used to describe data and explain relation between one dependent and independent variable.

1. Binary Losgistic: predict the probability of binary outcome. yes/no , true/false...

2. Multinominal Logisitic: predict the probabiltiy of one of three or more possible outcomes.

3. Ordinal Logisitic: predict the probability of an outcome that falls into predetermined order.

**sigmoid function:** math func to map predicted vaues to probabilities , maps real value into another vale witin a range of 0-1

concept of thrshold value: above threshold =1 , below threshold = 0



In [1]:
import sklearn
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
LR = LogisticRegression()
model = LR.fit(x_train,y_train)
pred = model.predict(x_test)
trainacc = LR.score(x_train,y_train)
testacc = accuracy_score(y_test,pred)

#coeff
print(model.coef_)
#slope
print(model.intercept_)

#multinomial regression
MLR = LogisticRegression(multi_class = 'multinomial')

#solver in logisitic
'''
small-med = liblinear
high-dimension = lbfgs , newton-cg
large data= sag, saga
'''
model1 = LogisticRegression(solver = 'newton-cg')

#penalty in logisctic
#l1 and l2 and elasticnet regularization
model = LogisticRegression(penalty='l1',solver='liblinear')
model = LogisticRegression(penalty='l2',solver='liblinear')
model = LogisticRegression(penalty='elasticnet',solver='liblinear')



**Artificial Neural Network**
1. Perceptron: building block and basic computational unit.
linear classifier , takes binary inputs and gives binary output.

2. Inputs and Weights: each input has its assigned weight which determines the influence of that input on output.

3. weighted sum of inputs and weight is calculated.

4. activation function: weighted sum is passed thru the activaton function , where the result is comapred to threshod an then he output is generated on its basis.

5. 1 if sum >= 0,
   0 otherwise.

they cannot learn complex patterns of relations, form the basis for MLPs.

**Neural Network:** neurons organised into layers.

1. neurons: basic computational unit , recives one or more input, performs a computation and produces an output.
2. Layers:
input layer:


In [ ]:
#input layer with 1 hiddenlayer:
model.add(Dense(units=20,input_dim=20,activation='relu'))
# 2ns hidden layer
model.add(Dense(units=10,activation='relu'))
#output layer
model.add(Dense(units=1,activation='sigmoid'))

#compilation
model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])
#trainin
model.fit(X_train,y_train,epochs=10,batch_size=32,validation_data=(X_test,y_test))


#### binary classification
1. sigmoid: for very large or very small inputs , gradient=0.

use 'Binary Cross-Entropy' for sigmoid.

2. Tanh: vanishing gradient , better than sigmoid since output is zero-centered.

#### hidden layer
3. Relu: +ve side , gradient =1.
-ve side, gradient =0 , neurons die use "Leaky Relu" or "Parametric Relu"

####multi class classification
4. Softmax: outputs not in hidden layers. use "log-softmax" for very large values.

use 'Categorical Cross-Entropy' as loss function.

In [ ]:
# leaky relu

from tensorflow.keras.layers import Dense,LeakyReLU
model = Sequential([
    Dense(64,activation=None , input_shape=(20,)),
    LeakyReLU(alpha=0.01),
    Dense(32,activation=None),
    LeakyReLU(alpha=0.01),
    Dense(1)
])

In [74]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


In [75]:
df = pd.read_csv("/content/heart.csv")
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [76]:
df.describe()

age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

In [78]:
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [79]:
X = df.drop('target', axis=1)
y = df['target']

X = pd.get_dummies(X, drop_first=True)


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print(X_train.shape, X_test.shape)

assert X_train.shape[1] == X_test.shape[1]


scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

(820, 13) (205, 13)


In [82]:
l1 = LogisticRegression(penalty='l1',solver='liblinear')
l1.fit(X_train_scaled,y_train)
l1_pred = l1.predict(X_test_scaled)

train_acc = l1.score(X_train_scaled,y_train)
test_acc = accuracy_score(y_test,l1_pred)

print(train_acc)
print(test_acc)

0.8719512195121951
0.7951219512195122


In [85]:
l2 = LogisticRegression(penalty='l2',solver='lbfgs')
l2.fit(X_train_scaled,y_train)
l2_pred = l2.predict(X_test_scaled)

train_acc = l2.score(X_train_scaled,y_train)
test_acc = accuracy_score(y_test,l2_pred)

print(train_acc)
print(test_acc)

0.8719512195121951
0.7951219512195122


In [87]:
l3 = LogisticRegression(penalty='elasticnet',solver='saga',l1_ratio=0.5)
l3.fit(X_train_scaled,y_train)
l3_pred = l3.predict(X_test_scaled)

train_acc = l3.score(X_train_scaled,y_train)
test_acc = accuracy_score(y_test,l3_pred)

print(train_acc)
print(test_acc)

0.8719512195121951
0.7951219512195122


While implementing L1 and Elastic Net penalties in Logistic Regression, I encountered solver-related errors.
Specifically:

L1 penalty does not work with the default solver (lbfgs), so an error appears:
“ValueError: Solver lbfgs does not support L1 penalty”
To fix this, I changed the solver to liblinear .
Elastic Net only works with solver='saga'.
Attempting to use any other solver results in the error:
“ValueError: l1_ratio is only supported for saga solver with elasticnet penalty”
Additionally, Elastic Net requires an extra parameter:
l1_ratio
l1_ratio = 1 → behaves like L1
l1_ratio = 0 → behaves like L2
0 < l1_ratio < 1 → mix of both

Other parameter changes:
Scaling was required because Logistic Regression performs better when features are normalize


In [88]:
# task 2
df2 = pd.read_csv("/content/iris.csv")
df2.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [91]:
df2['species']=df2['species'].astype('category').cat.codes

In [93]:
df2.dtypes

sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
species            int8
dtype: object

In [94]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    int8   
dtypes: float64(4), int8(1)
memory usage: 5.0 KB


In [95]:
df2.describe()

sepal_length  sepal_width  petal_length  petal_width     species
count    150.000000   150.000000    150.000000   150.000000  150.000000
mean       5.843333     3.054000      3.758667     1.198667    1.000000
std        0.828066     0.433594      1.764420     0.763161    0.819232
min        4.300000     2.000000      1.000000     0.100000    0.000000
25%        5.100000     2.800000      1.600000     0.300000    0.000000
50%        5.800000     3.000000      4.350000     1.300000    1.000000
75%        6.400000     3.300000      5.100000     1.800000    2.000000
max        7.900000     4.400000      6.900000     2.500000    2.000000

In [97]:
df2.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

In [110]:
X2 = df2.drop('species',axis=1)
y2=df2['species']

In [113]:
X2_train,X2_test,y2_train,y2_test = train_test_split(X2,y2,test_size=0.2,random_state=42,stratify=y)

In [114]:
print(X2_train.shape, X2_test.shape)


assert isinstance(X2_train, pd.DataFrame)
assert isinstance(X2_test, pd.DataFrame)
assert X2_train.shape[1] == X2_test.shape[1]   # ✅ This will PASS


(120, 4) (30, 4)


In [116]:
scaler = StandardScaler()
X2_train_scaled = scaler.fit_transform(X2_train)
X2_test_scaled = scaler.transform(X2_test)

In [119]:
solvers = ['lbfgs', 'liblinear', 'newton-cg', 'newton-cholesky', 'sag', 'saga']

results = []

for solver in solvers:
    try:
        model = LogisticRegression(solver=solver)
        model.fit(X2_train_scaled, y2_train)

        train_acc = accuracy_score(y2_train, model.predict(X2_train_scaled))
        test_acc = accuracy_score(y2_test, model.predict(X2_test_scaled))

        results.append([solver, train_acc, test_acc])

    except Exception as e:
        results.append([solver, "Error", "Error"])
df_results_iris = pd.DataFrame(
    results,
    columns=["Solver", "Training Accuracy", "Testing Accuracy"]
)
df_results_iris

Solver  Training Accuracy  Testing Accuracy
0            lbfgs           0.958333          0.933333
1        liblinear           0.925000          0.833333
2        newton-cg           0.958333          0.933333
3  newton-cholesky           0.958333          0.933333
4              sag           0.958333          0.933333
5             saga           0.958333          0.933333

all solvers performed extremely well on the Iris dataset because:
The dataset is small (150 samples)
Features are clean and well-separated
Problem is multi-class but simple
Effect of solvers
Solver	Behavior
lbfgs	Stable, fast, performs best for small/medium multiclass problems.
newton-cg	Similar to lbfgs, high accuracy, strong for multiclass.
newton-cholesky	Very fast convergence, stable like newton-cg.
liblinear	Designed for binary classification; still works but slower and less ideal for multiclass.
sag	Good for large datasets but less stable on very small datasets.
saga	Works with all penalties, handles large datasets; slightly slower here.
Matching sklearn documentation?
Yes — sklearn says:
lbfgs, newton-cg, newton-cholesky → best for multiclass and small/medium datasets
liblinear → best for binary, not ideal for large or multiclass
sag/saga → best for large datasets
This matches our observations: the large-dataset solvers (sag, saga) did not outperform lbfgs on Iris because the dataset is too small.
Best solver for Iris? Why?
lbfgs or newton-cholesky
Fast convergence
Stable
Performs best on small to medium-sized multiclass problems

In [120]:
import pandas as pd

df = pd.read_csv("heart.csv")   # Kaggle heart disease dataset

X = df.drop("target", axis=1)
y = df["target"]
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
solvers = ['lbfgs', 'liblinear', 'newton-cg', 'newton-cholesky', 'sag', 'saga']

results_heart = []

for solver in solvers:
    try:
        model = LogisticRegression(solver=solver, max_iter=3000)
        model.fit(X_train_scaled, y_train)

        train_acc = accuracy_score(y_train, model.predict(X_train_scaled))
        test_acc = accuracy_score(y_test, model.predict(X_test_scaled))

        results_heart.append([solver, train_acc, test_acc])

    except Exception as e:
        results_heart.append([solver, "Error", "Error"])
df_results_heart = pd.DataFrame(
    results_heart,
    columns=["Solver", "Training Accuracy", "Testing Accuracy"]
)
df_results_heart

Solver  Training Accuracy  Testing Accuracy
0            lbfgs           0.871951          0.795122
1        liblinear           0.871951          0.795122
2        newton-cg           0.871951          0.795122
3  newton-cholesky           0.871951          0.795122
4              sag           0.871951          0.795122
5             saga           0.871951          0.795122

size is not the main reason solvers behave differently.
What matters more is:
How difficult the problem is
Whether it’s binary (2 classes) or multiclass (3+ classes)

heart - all perform equally well since binary class

iris - liblinear performs poorly since data is multiclass

In [121]:
# task 3
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression, Perceptron
from sklearn.metrics import accuracy_score
import pandas as pd

iris = pd.read_csv("/content/iris.csv")
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [122]:
iris['species'] = iris['species'].astype('category').cat.codes

In [125]:
X = iris.drop('species',axis=1)
y = iris['species']

In [126]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [127]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [133]:
lr = LogisticRegression()
lr.fit(X_train_scaled,y_train)

lr_train_acc = accuracy_score(y_train,lr.predict(X_train_scaled))
lr_test_acc = accuracy_score(y_test,lr.predict(X_test_scaled))

print(f"lr train acc: {lr_train_acc:.2f}")
print(f"lr test acc: {lr_test_acc:.2f}")

lr train acc: 0.97
lr test acc: 1.00


In [134]:
per = Perceptron()
per.fit(X_train_scaled,y_train)

per_train_acc = accuracy_score(y_train,per.predict(X_train_scaled))
per_test_acc = accuracy_score(y_test,per.predict(X_test))

print(f"per train acc: {per_train_acc:.2f}")
print(f"per test acc: {per_test_acc:.2f}")

per train acc: 0.88
per test acc: 0.37


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but Perceptron was fitted without feature names
  warnings.warn(


In [135]:
results = pd.DataFrame({
    "Model": ["Logistic Regression", "Perceptron"],
    "Training Accuracy": [lr_train_acc, per_train_acc],
    "Testing Accuracy": [lr_test_acc, per_test_acc]
})

results


Model  Training Accuracy  Testing Accuracy
0  Logistic Regression           0.966667          1.000000
1           Perceptron           0.883333          0.366667

Model	Behavior
Logistic Regression	Usually achieves higher & more stable accuracy.
Perceptron	Slightly lower accuracy and more variance due to being a harder-line classifier.
Logistic Regression tends to:
Converge better
Generalize more smoothly
Produce probabilistic outputs
The Perceptron, being a linear classifier using a hard threshold, is very sensitive to:
Learning rate
Feature scaling
Data not being perfectly linearly separable
Thus, Logistic Regression almost always performs more reliably on the Iris dataset.
"""